lib/reports/BackupReportManager.js

```diff
@@ -0,0 +1,224 @@
+'use strict';
+
+const _ = require('lodash');
+const Promise = require('bluebird');
+const moment = require('moment-timezone');
+const Repository = require('../db').Repository;
+const CONST = require('../constants');
+const config = require('../config');
+class BackupReportManager {
+  static createInstanceBackupRecord(instance_id, start_time, end_time) {
+    return Promise
+      .all([
+        this.getBackupTriggerRecord(instance_id, start_time, end_time),
+        this.getBackupResult(instance_id, start_time, end_time),
+        this.getInstanceDeleteTime(instance_id),
+        this.getInstanceCreateTime(instance_id)
+      ])
+      .spread((triggeredBackups, resultBackups, instanceDeleteTime, instanceCreateTime) => {
+        const instance_record = {};
+        const noBackupDays = [];
+        //start index in resultBackuplist
+        let start_index = 0;
+        let threeBackupFailureCount = 0;
+        let currentDay = new Date(start_time);
+        let lastFailDay;
+        let secondlastFailDay;
+        let BackupSuccessCount = 0;
+        if (instanceCreateTime) {
+          if (start_time < moment.utc(instanceCreateTime.createdAt).startOf('day').toDate()) {
+            currentDay = moment.utc(instanceCreateTime.createdAt).startOf('day').toDate();
+          }
+          _.assign(instance_record, {
+            instanceCreateTime: instanceCreateTime.createdAt
+          });
+        } else {
+          currentDay = new Date(end_time);
+        }
+        if (instanceDeleteTime && (end_time > instanceDeleteTime.createdAt)) {
+          _.assign(instance_record, {
+            instanceDeleteTime: instanceDeleteTime.createdAt
+          });
+          end_time = instanceDeleteTime.createdAt;
+        }
+        for (; currentDay < end_time; currentDay = moment(currentDay).add(1, 'days').toDate()) {
```

Comment:
A small comment on the compute logic below helps. Its difficult to decipher whats happening in the below loops.